WalletWasabi/Stores/SqliteStorage.cs

```diff
@@ -0,0 +1,241 @@
+using Microsoft.Data.Sqlite;
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using WalletWasabi.Backend.Models;
+
+namespace WalletWasabi.Stores;
+
+/// <summary>
+/// Sqlite-based storage for block filters.
+/// </summary>
+/// <seealso href="https://learn.microsoft.com/en-gb/dotnet/standard/data/sqlite/async">
+/// Async is not supported at the moment, so no async methods were used in this implementation.
+/// </seealso>
+public class SqliteStorage : IDisposable
+{
+	private bool _disposedValue;
+
+	private SqliteStorage(SqliteConnection connection)
+	{
+		Connection = connection;
+	}
+
+	private SqliteConnection Connection { get; }
```

Comment:
Usually connections are not stored in fields and properties, but instantiated on demand. Internally there's a pool and an inner connection object (one more controversial thing in ADO.NET), so having a connection string cached in a field would be a better solution.

Sqlite is a bit different because it's a in-process RDBMS, so it can be kept as is.

Reply:
Disposal of an ADO.NET connection may not dispose the underlying physical connection due to pooling turned on under the hood (SQLite isn't an exception and does it by default, [see code](https://github.com/dotnet/efcore/blob/bdd9846218b002005321efed1cf5195cae12f1f2/src/Microsoft.Data.Sqlite.Core/SqliteConnectionStringBuilder.cs#L57)). So, without opting out the code will continue to work exactly like it is now, but it will follow guidelines of ADO.NET usage and free resources when they are not used for some time ([source](https://github.com/dotnet/efcore/blob/bdd9846218b002005321efed1cf5195cae12f1f2/src/Microsoft.Data.Sqlite.Core/SqliteConnectionPool.cs#L30).

```
// Makes a new logical connection.
using var connection = new SqliteConnection();

// Gets an existing physical connection from the pool or opens it if it doesn't exist yet.
connection.Open();

// Dispose just returns the physical connection to the pool.
```

There's no performance penalties since you don't opt out from pooling (:

That design is weird, but it was made to improve performance of applications which have been made before pooling got its place in ADO.NET without changing these applications.

Reply:
In case of SQLite it's fine to keep it as is (it is an in process database), but the general practice is disposal of a connection right after finishing all commands in the method. Why so? There are two reasons:

1. A physical connection can be shared thankfully for pooling. That's useful when you have more than one place accessing the same database and potentially from different threads surely if both of them are not accessing it simultaneously.

2. For out of process databases there's a limited number of connections, so a better option here is to limit physical connections made by clients. A good example is PostgreSQL which starts an instance for each incoming connection and stops it when the connection closes. It's a very expensive operation here. Moreover, the listening server has only 100 parallel connections by default.

Sure, we won't use any out of process database, but you said that you are going to use SQLite in another place, so reason number one still applies even if it's partly valid due to the nature of SQLite.

Reply:
Oh, forgot to mention a primary reason behind having a new connection each time. The server can terminate it, so with a single open the service is screwed. Not sure how it applies to SQLite since there's no server, but I have no idea about it internals. Therefore, if a native connection referenced by a handle held can be invalidated for whatever reason, then that reason is valid too.

Honestly, I'm not aware about SQLite internals and wouldn't rely on its implementation. That's why I highly recommend getting a new connection each time.

Reply:
I'm okay with it.

I asked a question in Slack about locks and possibility to remove them due to having SQLite transactions. Perhaps, you have to instantiate a new connection per each method call of `SqliteStorage` since no parallel commands are allowed per connection in ADO.NET and databases I have used (except SQL Server with MARS enabled, but they still run synchronously). Maybe it's different with SQLite, no idea about that (: